src/VisualStudio/InteractiveServices/Interactive/VsInteractiveWindowProvider.cs

```diff
@@ -114,11 +117,55 @@ public IVsInteractiveWindow Open(int instanceId, bool focus)
             return _vsInteractiveWindow;
         }
 
-        private static ImmutableArray<IInteractiveWindowCommand> FilterCommands(IInteractiveWindowCommand[] commands, string contentType)
+        private static ImmutableArray<IInteractiveWindowCommand> GetApplicableCommands(IInteractiveWindowCommand[] commands, string coreContentType, string specializedContentType)
         {
-            return commands.Where(
+            // get all commands of coreContentType - generic interactive window commands
+            var interactiveCommands = commands.Where(
                 c => c.GetType().GetCustomAttributes(typeof(ContentTypeAttribute), inherit: true).Any(
-                    a => ((ContentTypeAttribute)a).ContentTypes == contentType)).ToImmutableArray();
-        }
+                    a => ((ContentTypeAttribute)a).ContentTypes == coreContentType)).ToArray();
+
+            // get all commands of specializedContentType - smart C#/VB command implementations
+            var specializedInteractiveCommands = commands.Where(
+                c => c.GetType().GetCustomAttributes(typeof(ContentTypeAttribute), inherit: true).Any(
+                    a => ((ContentTypeAttribute)a).ContentTypes == specializedContentType)).ToArray();
+
+            // We should choose specialized C#/VB commands over generic core interactive window commands
+            // so we replace the generic commands with specialized commands if both exist.
+            // Command can have multiple names. We need to compare every name to find match.
+            // Build a map of names and associated command first
+
+            Dictionary<string, int> interactiveCommandMap = new Dictionary<string, int>();
+            Dictionary<string, int> specializedInteractiveCommandMap = new Dictionary<string, int>();
+
+            for (int i = 0; i < interactiveCommands.Length; i++)
+            {
+                foreach (var name in interactiveCommands[i].Names)
+                        interactiveCommandMap.Add(name, i);
+            }
+
+            for (int i = 0; i < specializedInteractiveCommands.Length; i++)
+            {
+                foreach (var name in specializedInteractiveCommands[i].Names)
+                    specializedInteractiveCommandMap.Add(name, i);
+            }
+
+            //swap out generic commands with the specialized version
+            bool matchFound = false;
+            foreach ( var entry1 in interactiveCommandMap)
+            {
+                foreach( var entry2 in specializedInteractiveCommandMap)
+                {
+                    if (string.Equals(entry1.Key, entry2.Key, StringComparison.Ordinal))
+                    {
+                        Debug.Assert(matchFound == false, "Found more than one specialized commands with same name");
+                        interactiveCommands[entry1.Value] = specializedInteractiveCommands[entry2.Value];
```

Comment:
This replaces all names for the original command. Is that intended? (If the base command has names A and B and the specialized command has name A only, should the name B bind to the original command or no command?)